scipy/interpolate/_polyint.py

```diff
@@ -683,8 +758,89 @@ def _evaluate(self, x):
                 p[r[:-1]] = self.yi[r[-1]]
         return p
 
+    def derivative(self, x, der=1):
+        """
+        Evaluate a single derivative of the polynomial at the point x.
+
+        Parameters
+        ----------
+        x : array_like
+            Point or points at which to evaluate the derivatives
+        der : integer, optional
+            Which derivative to evaluate (default: first derivative).
+            This number includes the function value as 0th derivative.
+
+        Returns
+        -------
+        d : ndarray
+            Derivative interpolated at the x-points. Shape of `d` is
+            determined by replacing the interpolation axis in the
+            original array with the shape of `x`.
+        """
+        x, x_shape = self._prepare_x(x)
+        y = self._evaluate_derivatives(x, der+1, all_lower=False)
+        return self._finish_y(y, x_shape)
+
+    def _evaluate_derivatives(self, x, der=None, all_lower=True):
+        # NB: der here is not the order of the highest derivative;
+        # instead, it is the size of the derivatives matrix that
+        # would be returned with all_lower=True, including the
+        # '0th' derivative (the undifferentiated function).
+        # E.g. to evaluate the 5th derivative alone, call
+        # _evaluate_derivatives(x, der=6, all_lower=False).
+
+        if (not all_lower) and (x.size == 0 or self.r == 0):
+            return np.zeros((0, self.r), dtype=self.dtype)
+
+        if (not all_lower) and der == 1:
+            return self._evaluate(x)
+
+        if (not all_lower) and (der > self.n):
+            return np.zeros((len(x), self.r), dtype=self.dtype)
 
-def barycentric_interpolate(xi, yi, x, axis=0):
+        if der is None:
+            der = self.n
+
+        if all_lower and (x.size == 0 or self.r == 0):
+            return np.zeros((der, len(x), self.r), dtype=self.dtype)
+
+        if self._diff_cij is None:
+            # c[i,j] = xi[i] - xi[j]
+            c = self.xi[..., np.newaxis] - self.xi
+            # c[i,j] = (w[j] / w[i]) / (xi[i] - xi[j])
+            c = np.divide(self.wi, c * self.wi[..., np.newaxis],
+                          out=np.zeros_like(c), where=c != 0)
```

Comment:
Not sure what is going on here. `out=np.zeros_like(c)` signals "throw away the result", but then it is returned, ouch. It is probably deliberately ingenious (which it is!), but can we please write it in a more boring way?

Reply:
This was one of the changes I made to avoid rewriting the diagonal back a forth (as c is 0 along the diagonal). Here the division is performed normally apart from where `c==0` where the result is instead left unchanged in the `out` array (hence is zero as desired).

Would you be happy with a comment explaining what is happening?

Reply:
Unfortunately I haven't had any time to keep up with this lately so I'm out of the loop, but I'd actually be a bit surprised if the new version with its quadratic comparisons was faster than the previous version with two linear fills. 

If it's a meaningful difference it's probably worth propagating to the older barycentric interpolation code which uses the same two diagonal fills.

@j-bowhay I guess you've checked the new version is faster - do you have some numbers?

Reply:
I have to admit I thought I benchmarked this but clearly not. There is no applicable difference so I will revert the change:
```
import numpy as np
import perfplot

def setup(n):
    xi = np.random.random(n)
    c =  xi - xi[:, np.newaxis]
    np.fill_diagonal(c, 0)
    return c

def old(x):
    c = np.copy(x)
    np.fill_diagonal(c, 1)
    c  = 1/c
    np.fill_diagonal(c,0)
    return c

def new(x):
    c = np.copy(x)
    return np.divide(1, c, out=np.zeros_like(c), where=c!=0)

perfplot.show(setup=setup, kernels=[old, new], n_range=[2**k for k in range(15)])
"""
![image](https://github.com/scipy/scipy/assets/60778417/518852e4-8a97-4751-8497-62dc4497b076)

Reply:
I'm fine with just adding a comment, FTR.